### DISCUSSION TITLE
Gradual typing

### DISCUSSION THREAD

**Discussion by jryans** [jank-lang/jank#76] [Issue Triage] [answered]
I am curious what you have planned in terms of gradual typing. I saw it mentioned on the homepage, but I haven't been able to find much more detail. Maybe it's still all up in the air, but even still, I'm curious how you are thinking about it.

Thanks for working on this project! 😄

**Comment by jeaye** [comment, marked as answer]
Gradual typing is entirely up in the air right now. But my general plan, once I have parity with Clojure and the performance, portability, and usability is in a good state, is to dig into compiler optimizations which can be made by static analysis.

People claim that Clojure can't do this, since it's such a dynamic language, BUT clj-kondo does it all the time. We all know the expected inputs to `clojure.core` functions and we build basically all of our programs on top of those functions. clj-kondo can tell me when I'm giving a function invalid arguments, so why can't the compiler? Any why can't we monomorphize to optimize for specific types where applicable? And why can't we take our specs (or malli schemas) and do static analysis based on that (spoiler: we can; clj-kondo does this)?

The follow-on question is: if clj-kondo is already doing it, why bother? Well, I'm not just interested in catching errors. I'm interested in optimizing code for scenarios when we have type info. If the compiler can know that `+` is always going to take in a `real` or an `integer`, it can monomorphize those into C++ overloads and support unboxed arities by default. Then we get allocation-free arithmetic without me having to hard-code things like this in.

So, at this point, I'm less focused on building a robust gradual type system specifically and more on the just building a smarter compiler which uses the information available to do a better job. My goal is for this to happen at zero additional cost. Clearly we've seen with Typed Clojure that annotating everything has a high cost. That's not what I'm thinking. Instead, the basis of my reasoning will be "How much can we do with absolutely zero cost?"

I think the answer is "A lot more than Clojure is currently doing."

Thanks for the interest! It'll be another year, at least, before I'm tackling these things, if I had to guess. I'm looking forward to it, though.
